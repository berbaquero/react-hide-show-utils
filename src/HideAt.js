import React from 'react';
import PropTypes from 'prop-types';

/**
 * Descriptively hide children components, based on a breakpoint
 * Uses matchMedia
 * Client-side only
 */

class HideAt extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      visible: false,
    };
    this.mql = window.matchMedia(props.breakpoint);
    this.mql.addListener(this.updateVisibility.bind(this));
  }

  updateVisibility() {
    const shouldHide = this.mql.matches;
    this.setState({
      visible: shouldHide,
    });
  }

  // Check visibility before mounting (ergo `render()`-ing)
  // so it does not render `children` if the breakpoint is already active
  componentWillMount() {
    this.updateVisibility();
  }

  render() {
    return this.state.visible ? this.props.children : null;
  }
}

HideAt.propTypes = {
  breakpoint: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired,
};

export default HideAt;
