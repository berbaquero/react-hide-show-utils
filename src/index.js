import React from 'react';
import PropTypes from 'prop-types';

/**
 * Descriptively hide or show children components, based on a breakpoint
 * Uses matchMedia
 * Client-side only
 */

class HideShow extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      visible: false,
    };
    this.mql = window.matchMedia(props.breakpoint);
    this.mql.addListener(this.updateVisibility.bind(this));
  }

  updateVisibility() {
    const breakpointActive = this.mql.matches;
    if (this.props.hide) {
      this.setState({ visible: !breakpointActive });
    } else {
      this.setState({ visible: breakpointActive });
    }
  }

  // Check visibility before mounting (ergo `render()`-ing)
  // so it does not render `children` if the breakpoint is already active
  componentWillMount() {
    this.updateVisibility();
  }

  render() {
    return this.state.visible ? this.props.children : null;
  }
}

const commonPropTypes = {
  breakpoint: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired,
};

HideShow.propTypes = {
  hide: PropTypes.bool.isRequired,
};

const HideAt = ({ breakpoint, children }) => (
  <HideShow breakpoint={breakpoint} hide={true}>
    {children}
  </HideShow>
);

HideAt.propTypes = commonPropTypes;

const ShowAt = ({ breakpoint, children }) => (
  <HideShow breakpoint={breakpoint} hide={false}>
    {children}
  </HideShow>
);

ShowAt.propTypes = commonPropTypes;

export { HideAt, ShowAt };
