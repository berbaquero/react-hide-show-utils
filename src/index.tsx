import * as React from 'react';

interface HideShowProps {
  breakpoint: string;
  hide: boolean;
}

interface HideShowState {
  visible: boolean;
}

/**
 * Descriptively hide or show children components, based on a breakpoint.
 * Uses matchMedia.
 */
class HideShow extends React.Component<HideShowProps, HideShowState> {
  private mql?: MediaQueryList;
  private isClient: boolean;
  constructor(props: HideShowProps) {
    super(props);
    this.isClient = typeof window !== 'undefined';
    this.updateVisibility = this.updateVisibility.bind(this);
    if (this.isClient) {
      // Only initialize matchMedia on the client
      this.mql = window.matchMedia(props.breakpoint);
      this.mql.addListener(this.updateVisibility);
    }
    // Set initial visibility for first render
    this.state = {
      visible: this.isClient ? isVisible(props.hide, this.mql!.matches) : false,
    };
  }

  updateVisibility() {
    if (!this.isClient) return; // Only on the client
    this.setState({ visible: isVisible(this.props.hide, this.mql!.matches) });
  }

  componentWillUnmount() {
    if (this.mql) {
      this.mql.removeListener(this.updateVisibility);
    }
  }

  render() {
    return this.state.visible ? this.props.children : null;
  }
}

export interface HideShowAtProps {
  breakpoint: HideShowProps['breakpoint'];
  children?: React.ReactNode;
}

const HideAt: React.FC<HideShowAtProps> = function({ breakpoint, children }) {
  return (
    <HideShow breakpoint={breakpoint} hide={true}>
      {children}
    </HideShow>
  );
};

const ShowAt: React.FC<HideShowAtProps> = function({ breakpoint, children }) {
  return (
    <HideShow breakpoint={breakpoint} hide={false}>
      {children}
    </HideShow>
  );
};

export { HideAt, ShowAt };

function isVisible(isHide: boolean, mqIsActive: boolean): boolean {
  return isHide ? !mqIsActive : mqIsActive;
}
