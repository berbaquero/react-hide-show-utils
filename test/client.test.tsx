import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { HideAt, ShowAt } from '../src/index';

// Mock
window.matchMedia = jest.fn().mockImplementation(query => {
  return {
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
  };
});

describe('<HideAt>', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <HideAt breakpoint="(min-width: 10em)">
      <h1>Anything here</h1>
    </HideAt>,
    div
  );

  it('should render its children', () => {
    expect(div).toMatchSnapshot();
  });
});

describe('<ShowAt>', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <ShowAt breakpoint="(min-width: 10em)">
      <h1>Anything here</h1>
    </ShowAt>,
    div
  );

  it('should not render its children', () => {
    expect(div).toMatchSnapshot();
  });
});
