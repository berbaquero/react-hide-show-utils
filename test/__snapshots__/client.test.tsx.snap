// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HideAt> should render its children 1`] = `
<div>
  <h1>
    Anything here
  </h1>
</div>
`;

exports[`<ShowAt> should not render its children 1`] = `<div />`;
